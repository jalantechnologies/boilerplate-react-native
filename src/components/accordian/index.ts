export { default as Accordion } from './accordian';
export { default as AccordionItem } from './accordianItem';
export type { AccordionProps, AccordionItemProps } from './types';
