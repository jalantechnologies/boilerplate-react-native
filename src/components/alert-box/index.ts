export { AlertProvider, useAlert } from './AlertProvider';
export { AlertType } from '../../types/alert';
