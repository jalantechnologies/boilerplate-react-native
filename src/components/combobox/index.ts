export { ComboboxInput } from './comboboxInput';
export { ComboboxOptionList } from './comboboxOptionList';
export { ComboboxOptionItem } from './comboboxOptionItem';
export * from './hooks/useCombobox';
export * from './types';
export * from './styles';
