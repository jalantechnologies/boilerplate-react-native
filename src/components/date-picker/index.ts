export { DatePicker } from './DatePicker';
export * from '../../types/date-picker';
