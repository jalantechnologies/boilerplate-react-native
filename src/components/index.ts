export { default as Brand } from './brand/brand';
export { default as ErrorFallback } from './error-fallback/error-fallback';
export { default as OTPInput } from './otp-input/otp-input';
