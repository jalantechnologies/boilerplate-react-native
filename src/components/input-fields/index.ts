export { default as Input } from './input';
export type { InputProps, InputSize, InputVariant } from './types';
