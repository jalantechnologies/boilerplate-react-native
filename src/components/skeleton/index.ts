export { default as Skeleton } from './skeleton';
export type { SkeletonProps, SkeletonShape } from './types';
export {
  DEFAULT_SKELETON_SHAPE,
  DEFAULT_SKELETON_WIDTH,
  DEFAULT_SKELETON_HEIGHT,
  DEFAULT_ANIMATION_DURATION,
} from './constants';
