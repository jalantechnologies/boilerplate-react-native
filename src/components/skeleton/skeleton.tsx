import React, { useEffect, useRef } from 'react';
import { Animated } from 'react-native';

import {
  DEFAULT_SKELETON_SHAPE,
  DEFAULT_SKELETON_WIDTH,
  DEFAULT_SKELETON_HEIGHT,
  DEFAULT_ANIMATION_DURATION,
} from './constants';
import type { SkeletonProps } from './types';

/**
 * Skeleton Component - Displays a placeholder UI to indicate content loading.
 *
 * Example usage:
 * ```tsx
 * // Rectangle skeleton
 * <Skeleton isVisible={true} shape="rect" width={200} height={20} animation />
 *
 * // Circular skeleton
 * <Skeleton isVisible={true} shape="circle" width={50} animation />
 * ```
 *
 * Note: Styling will be enhanced with NativeWind in the future.
 */
const Skeleton: React.FC<SkeletonProps> = ({
  isVisible = true,
  shape = DEFAULT_SKELETON_SHAPE,
  width = DEFAULT_SKELETON_WIDTH,
  height = DEFAULT_SKELETON_HEIGHT,
  animation = true,
  style,
}) => {
  if (!isVisible) {
    return null;
  }

  const opacityAnim = useRef(new Animated.Value(0.5)).current;

  useEffect(() => {
    if (animation) {
      Animated.loop(
        Animated.sequence([
          Animated.timing(opacityAnim, {
            toValue: 1,
            duration: DEFAULT_ANIMATION_DURATION,
            useNativeDriver: true,
          }),
          Animated.timing(opacityAnim, {
            toValue: 0.5,
            duration: DEFAULT_ANIMATION_DURATION,
            useNativeDriver: true,
          }),
        ]),
      ).start();
    }
  }, [animation, opacityAnim]);

  const animatedStyle = animation ? { opacity: opacityAnim } : {};

  // For circles, use width for both dimensions; for rectangles, use provided width and height.
  const containerStyle = {
    width,
    height: shape === 'circle' ? width : height,
    borderRadius: shape === 'circle' ? width / 2 : 4,
    backgroundColor: '#e0e0e0',
  };

  return <Animated.View style={[containerStyle, animatedStyle, style]} />;
};

export default Skeleton;
