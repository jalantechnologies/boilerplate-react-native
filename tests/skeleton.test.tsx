import { render } from '@testing-library/react-native';
import React from 'react';

import { Skeleton } from '../src/components/skeleton';
import {
  DEFAULT_SKELETON_WIDTH,
  DEFAULT_SKELETON_HEIGHT,
} from '../src/components/skeleton/constants';

describe('Skeleton Component', () => {
  test('renders correctly with default props', () => {
    const { toJSON } = render(<Skeleton isVisible={true} />);
    expect(toJSON()).toMatchSnapshot();
  });

  test('renders as a rectangle by default', () => {
    const { toJSON } = render(<Skeleton isVisible={true} />);
    const tree = toJSON();
    // For a rectangle, the container style should have:
    // width: DEFAULT_SKELETON_WIDTH,
    // height: DEFAULT_SKELETON_HEIGHT,
    // borderRadius: 4 (as defined in our component for non-circular shapes)
    const styleArray =
      tree && Array.isArray(tree)
        ? tree.flatMap(node =>
            Array.isArray(node.props.style) ? node.props.style : [node.props.style],
          )
        : tree && Array.isArray(tree.props.style)
        ? tree.props.style
        : tree
        ? [tree.props.style]
        : [];
    expect(styleArray).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          width: DEFAULT_SKELETON_WIDTH,
          height: DEFAULT_SKELETON_HEIGHT,
          borderRadius: 4,
        }),
      ]),
    );
  });

  test('renders as a circle when shape is "circle"', () => {
    const customWidth = 50;
    const { toJSON } = render(<Skeleton isVisible={true} shape="circle" width={customWidth} />);
    const tree = toJSON();
    // For a circle, height should equal width and borderRadius should be half of width.
    const styleArray =
      tree && Array.isArray(tree)
        ? tree.flatMap(node =>
            Array.isArray(node.props.style) ? node.props.style : [node.props.style],
          )
        : tree && Array.isArray(tree.props.style)
        ? tree.props.style
        : tree
        ? [tree.props.style]
        : [];
    expect(styleArray).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          width: customWidth,
          height: customWidth,
          borderRadius: customWidth / 2,
        }),
      ]),
    );
  });

  test('applies shimmering animation when enabled', () => {
    const { toJSON } = render(<Skeleton isVisible={true} animation />);
    const tree = toJSON();
    // Check if one of the style objects includes an "opacity" property.
    const styleArray =
      tree && !Array.isArray(tree) && tree.props && Array.isArray(tree.props.style)
        ? tree.props.style
        : tree && !Array.isArray(tree) && tree.props
        ? [tree.props.style]
        : [];
    const hasOpacity: boolean = styleArray.some(
      (style: { opacity?: number }) => style.opacity !== undefined,
    );
    expect(hasOpacity).toBe(true);
  });
});
