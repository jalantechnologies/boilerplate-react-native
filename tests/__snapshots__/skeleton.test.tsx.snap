// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Skeleton Component renders correctly with default props 1`] = `
<View
  collapsable={false}
  style={
    {
      "backgroundColor": "#e0e0e0",
      "borderRadius": 4,
      "height": 20,
      "opacity": 0.5,
      "width": 100,
    }
  }
/>
`;
